command-line/lib/Command/Key/Child.hs

```diff
@@ -56,34 +49,109 @@ mod liftCmd = command "child" $
             let ixs = castDerivationPath path
-            case foldM (deriveXPub scheme) xpub ixs of
+            case foldM (deriveXPub DerivationScheme2) xpub ixs of
+                Just child ->
+                    (,xpubToBytes child) <$> childHrpFor (fromEnum <$> ixs) hrp
                 Nothing ->
                     fail
                         "Couldn't derive child key. If you're trying to derive \
                         \children on a PUBLIC key, you must use soft indexes only."
-                Just child ->
-                    pure
-                        ( xpubToBytes child
-                        , fmap (const [humanReadablePart|xpub|] ) encoding
-                        )
 
-        Right xprv -> do
+        Right (hrp, xprv) -> do
             let ixs = castDerivationPath path
+            let scheme = if length ixs == 2 && hrp == CIP5.root_xsk
+                    then DerivationScheme1
+                    else DerivationScheme2
             let Identity child = foldM (\k -> pure . deriveXPrv scheme k) xprv ixs
-            pure
-                ( xprvToBytes child
-                , fmap (const [humanReadablePart|xprv|]) encoding
-                )
+            (,xprvToBytes child) <$> childHrpFor (fromEnum <$> ixs) hrp
+
+    hPutBytes stdout child (EBech32 hrp)
+  where
+    allowedPrefixes =
+        [ CIP5.root_xsk
+        , CIP5.acct_xsk
+        , CIP5.acct_xvk
+        ]
+
+    -- As a reminder, we really have two scenarios:
+    --
+    -- Byron Legacy:
+    --
+    --     m / rnd_account' / rnd_address'
+    --
+    --
+    -- Icarus & Shelley:
+    --
+    --     m / purpose' / coin_type' / account' / role / index
+    --
+    -- We do not allow derivations to anywhere in the path to avoid people
+    -- shooting themselves in the foot. Hence We only allow the following
+    -- transformations:
+    --
+    -- root_xsk => addr_xsk: (legacy)
+    --     m => m / rnd_account' / rnd_address
+    --
+    -- root_xsk => acct_xsk: (hard derivation from root to account)
+    --     m => m / purpose' / coin_type' / account'
+    --
+    -- root_xsk => acct_xsk: (hard derivation from root to address)
+    --     m => m / purpose' / coin_type' / account' / role / index
+    --
+    -- acct_xsk => addr_xsk: (hard derivation from account to address)
+    --     m / purpose' / coin_type' / account' => m / purpose' / coin_type' / account' / role / index
+    --
+    -- acct_xvk => addr_xvk: (soft derivation from account to address)
+    --     m / purpose' / coin_type' / account' => m / purpose' / coin_type' / account' / role / index
+    --
+    -- There's no use-case at the moment for accessing intermediate paths such
+    -- as m / purpose' or m / purpose' / coin_type' so we do not expose them.
+    childHrpFor [_,_,_,2,_] hrp
+        | hrp == CIP5.root_xsk = pure CIP5.stake_xsk
+
+    childHrpFor [_,_,_,3,_] hrp
+        | hrp == CIP5.root_xsk = pure CIP5.script_xsk
+
+    childHrpFor [_,_,_,4,_] hrp
```

Comment:
we allow rule=3 and rule=4 for scripts ... Why rule=3 is not enough?

Reply:
No, they translate to role=3 and role=4 from: https://hackmd.io/bdMljmoWR2yu4fqeAC7FCA